Documentation/ExtensionArchitecture/ConfigurationFiles/Index.rst

```diff
@@ -267,54 +273,87 @@ to be installed for this to work.
-  that would make the cached script concept break.
-- You must **never** use a :php:`use` statement in the files global scope -
-  that would make the cached script concept break and could conflict with other extensions.
-- The same goes for :php:`declare(strict_types=1)` and similar directives which must be placed
+-  You **MUST NOT** use a :php:`return` statement in the files global scope -
+   that would make the cached script concept break.
+
+-  You **MUST NOT** rely on the PHP constant :php:`__FILE__` for detection of
+   include path of the script - the configuration might be executed from
+   a cached script and therefore such information should be derived from
+   e.g. :php:`\TYPO3\CMS\Core\Utility\GeneralUtility::getFileAbsFileName()` or
+   :php:`ExtensionManagementUtility::extPath()`.
+
+
+-  You **MUST NOT** use :php:`use` inside :file:`ext_localconf.php` or :file:`ext_tables.php`.
+
+.. code-block:: diff
+
+   // do NOT use use:
+   -use TYPO3\CMS\Core\Resource\Security\FileMetadataPermissionsAspect;
+   -
+   -$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['t3lib/class.t3lib_tcemain.php']['processDatamapClass'][] = FileMetadataPermissionsAspect::class;
+   +$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['t3lib/class.t3lib_tcemain.php']['processDatamapClass'][] = \TYPO3\CMS\Core\Resource\Security\FileMetadataPermissionsAspect::class;
+
+- You **MUST NOT** use :php:`declare(strict_types=1)` and similar directives which must be placed
   at the very top of files: once all files of all extensions are merged, this condition is not
   fulfilled anymore leading to errors. So these must **never** be used here.
-- You should **not** rely on the PHP constant :php:`__FILE__` for detection of
-  include path of the script - the configuration might be executed from
-  a cached script and therefore such information should be derived from
-  e.g. :php:`\TYPO3\CMS\Core\Utility\GeneralUtility::getFileAbsFileName()` or
-  :php:`ExtensionManagementUtility::extPath()`.
 
-It is a good practice to use a directly called closure function to encapsulate all
-locally defined variables and thus keep them out of the surrounding scope. This
-avoids unexpected side-effects with files of other extensions.
+.. code-block:: diff
+
+   // do NOT use declare strict and other directives which MUST be placed at the top of the file
+   -declare(strict_types=1)
+
+
+-  You **MUST NOT** check for values of :php:`TYPO3_MODE` or :php:`TYPO3_REQUESTTYPE`
+   constants (e.g. :php:`if (TYPO3_MODE === 'BE')`) within these files as it limits the functionality
+   to cache the whole systems' configuration. Any extension author should remove the checks if not
+   explicitly necessary, and re-evaluate if these context-depending checks could go inside
+   the hooks / caller function directly., e.g. do not do::
+
+.. code-block:: diff
+
+   // do NOT do this:
+   -if (TYPO3_MODE === 'BE')
+
+-  You **SHOULD** check for the existence of the constants :php:`defined('TYPO3_MODE') or die();`
```

Comment:
I think we should be able to backport this to 10.4 and it might be better to make the change for the new TYPO3 constant as separate PR. Thanks for your comments, will push changes.